UT010029: Stream is closed with WebHook connector

## This is a...

<pre><code>
[ ] Feature request
[ ] Regression (a behavior that used to work and stopped working in a new release)
[b] Bug report  <!-- Please search GitHub for a similar issue or PR before submitting -->
[ ] Documentation issue or request
</code></pre>

## Description
WebHook connector reads the HTTP body as `InputStream` and replaces it with Syndesis specific JSON containing parameters/body. The binding (`org.apache.camel.http.common.DefaultHttpBinding`) that created the message (`org.apache.camel.http.common.HttpMessage`) reads from the `InputStream` of the `HttpServletRequest` and closes the stream. If the next processor in the exchange is a `org.apache.camel.processor.Pipeline` `org.apache.camel.util.ExchangeHelper::copyResults` is invoked and ends up calling `HttpMessage::createBody` that ends up calling `DefaultHttpBinding::parseBody` reading the `InputStream` for the second time.
At this time the `InputStream` is already closed so `UT010029: Stream is closed` is reported.

```
Caused by: java.io.IOException: UT010029: Stream is closed
	at io.undertow.servlet.spec.ServletInputStreamImpl.available(ServletInputStreamImpl.java:237) ~[undertow-servlet-1.4.25.Final.jar:1.4.25.Final]
	at org.apache.camel.http.common.HttpConverter.toInputStream(HttpConverter.java:86) ~[camel-http-common-2.21.1.jar:2.21.1]
	at org.apache.camel.http.common.HttpHelper.readRequestBodyFromServletRequest(HttpHelper.java:193) ~[camel-http-common-2.21.1.jar:2.21.1]
	at org.apache.camel.http.common.DefaultHttpBinding.parseBody(DefaultHttpBinding.java:577) ~[camel-http-common-2.21.1.jar:2.21.1]
	at org.apache.camel.http.common.HttpMessage.createBody(HttpMessage.java:78) ~[camel-http-common-2.21.1.jar:2.21.1]
```

--- a/app/connector/support/processor/src/main/java/io/syndesis/connector/support/processor/HttpRequestWrapperProcessor.java
+++ b/app/connector/support/processor/src/main/java/io/syndesis/connector/support/processor/HttpRequestWrapperProcessor.java
@@ -23,6 +23,8 @@ import io.syndesis.connector.support.processor.util.SimpleJsonSchemaInspector;
 import org.apache.camel.Exchange;
 import org.apache.camel.Message;
 import org.apache.camel.Processor;
+import org.apache.camel.impl.DefaultMessage;
+import org.apache.camel.util.ExchangeHelper;
 import org.apache.camel.util.ObjectHelper;
 
 import java.io.InputStream;
@@ -67,7 +69,11 @@ public class HttpRequestWrapperProcessor implements Processor {
             rootNode.putPOJO("body", body);
         }
 
-        message.setBody(Json.toString(rootNode));
+        final String newBody = Json.toString(rootNode);
+        final Message replacement = new DefaultMessage(exchange.getContext());
+        replacement.copyFromWithNewBody(message, newBody);
+
+        ExchangeHelper.replaceMessage(exchange, replacement, false);
     }
 
     public Set<String> getParameters() {

--- a/app/connector/support/processor/src/test/java/io/syndesis/connector/support/processor/WrapperProcessorTest.java
+++ b/app/connector/support/processor/src/test/java/io/syndesis/connector/support/processor/WrapperProcessorTest.java
@@ -15,23 +15,32 @@
  */
 package io.syndesis.connector.support.processor;
 
+import java.io.ByteArrayInputStream;
+import java.util.Arrays;
+import java.util.Collection;
+
 import com.fasterxml.jackson.databind.JsonNode;
 import com.fasterxml.jackson.databind.jsonFormatVisitors.JsonFormatTypes;
 import com.fasterxml.jackson.module.jsonSchema.JsonSchema;
 import com.fasterxml.jackson.module.jsonSchema.types.ObjectSchema;
+
 import io.syndesis.common.util.Json;
+
+import org.apache.camel.CamelContext;
 import org.apache.camel.Exchange;
 import org.apache.camel.Message;
+import org.apache.camel.spi.HeadersMapFactory;
 import org.junit.Test;
 import org.junit.runner.RunWith;
 import org.junit.runners.Parameterized;
 import org.junit.runners.Parameterized.Parameter;
 import org.junit.runners.Parameterized.Parameters;
-import org.mockito.Mockito;
+import org.mockito.ArgumentCaptor;
 
-import java.io.ByteArrayInputStream;
-import java.util.Arrays;
-import java.util.Collection;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
 
 @RunWith(Parameterized.class)
 public class WrapperProcessorTest {
@@ -77,16 +86,21 @@ public class WrapperProcessorTest {
         JsonNode schemaNode = Json.reader().forType(JsonNode.class).readTree(schemaStr);
         final HttpRequestWrapperProcessor processor = new HttpRequestWrapperProcessor(schemaNode);
 
-        final Exchange exchange = Mockito.mock(Exchange.class);
-        final Message message = Mockito.mock(Message.class);
-        Mockito.when(exchange.getIn()).thenReturn(message);
-        Mockito.when(message.getBody()).thenReturn(givenBody);
-        Mockito.when(message.getHeader("param1", String.class)).thenReturn("param_value1");
-        Mockito.when(message.getHeader("param2", String.class)).thenReturn("param_value2");
+        final Exchange exchange = mock(Exchange.class);
+        final Message message = mock(Message.class);
+        final CamelContext camelContext = mock(CamelContext.class);
+        when(camelContext.getHeadersMapFactory()).thenReturn(mock(HeadersMapFactory.class));
+        when(exchange.getIn()).thenReturn(message);
+        when(exchange.getContext()).thenReturn(camelContext);
+        when(message.getBody()).thenReturn(givenBody);
+        when(message.getHeader("param1", String.class)).thenReturn("param_value1");
+        when(message.getHeader("param2", String.class)).thenReturn("param_value2");
 
         processor.process(exchange);
 
-        Mockito.verify(message).setBody(replacedBody);
+        final ArgumentCaptor<Message> replacement = ArgumentCaptor.forClass(Message.class);
+        verify(exchange).setIn(replacement.capture());
+        assertThat(replacement.getValue().getBody()).isEqualTo(replacedBody);
     }
 
 }

